e2etests/tests/sessions.go

```diff
@@ -141,3 +144,142 @@ var _ = Describe("Router Host configuration", Ordered, func() {
 		})
 	})
 })
+
+var _ = Describe("Underlay BFD Configuration", Ordered, func() {
+	var cs clientset.Interface
+	routerPods := []*corev1.Pod{}
+	nodes := []corev1.Node{}
+
+	BeforeEach(func() {
+		cs = k8sclient.New()
+		var err error
+		routerPods, err = openperouter.RouterPods(cs)
+		Expect(err).NotTo(HaveOccurred())
+		nodesItems, err := cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		nodes = nodesItems.Items
+
+		// Enable BFD on leafkind
+		err = infra.UpdateLeafKindConfig(nodes, infra.EnableBFD)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		dumpIfFails(cs)
+		err := Updater.CleanAll()
+		Expect(err).NotTo(HaveOccurred())
+
+		By("waiting for router pods to rollout")
+		Eventually(func() error {
+			return openperouter.DaemonsetRolled(cs, routerPods)
+		}, 2*time.Minute, time.Second).ShouldNot(HaveOccurred())
+
+		err = infra.UpdateLeafKindConfig(nodes, !infra.EnableBFD)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	DescribeTable("should establish BFD sessions with the ToR",
+		func(underlay v1alpha1.Underlay) {
+			By("applying the underlay configuration")
+			err := Updater.Update(config.Resources{
+				Underlays: []v1alpha1.Underlay{
+					underlay,
+				},
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("validating BFD sessions are established")
+			exec := executor.ForContainer(infra.KindLeaf)
+			Eventually(func() error {
+				bfdPeers, err := frr.GetBFDPeers(exec)
+				if err != nil {
+					return err
+				}
+
+				if len(bfdPeers.Peers) != len(nodes) {
+					return fmt.Errorf("expecting %d BFD peers, got %d", len(nodes), len(bfdPeers.Peers))
+				}
+
+				for _, node := range nodes {
+					neighborIP, err := infra.NeighborIP(infra.KindLeaf, node.Name)
+					Expect(err).NotTo(HaveOccurred())
+
+					peer, ok := bfdPeers.Peers[neighborIP]
+					if !ok {
+						return fmt.Errorf("BFD session with %s not found", neighborIP)
+					}
+					if peer.Status != "up" {
+						return fmt.Errorf("BFD session with %s is not up", neighborIP)
+					}
+				}
+				return nil
+			}, 3*time.Minute, time.Second).ShouldNot(HaveOccurred())
+
+			By("validating BGP sessions are still established")
+			for _, node := range nodes {
+				neighborIP, err := infra.NeighborIP(infra.KindLeaf, node.Name)
+				Expect(err).NotTo(HaveOccurred())
+				validateSessionWithNeighbor(infra.KindLeaf, node.Name, exec, neighborIP, Established)
+			}
+
+			if underlay.Spec.Neighbors[0].BFD != nil && underlay.Spec.Neighbors[0].BFD.TransmitInterval != nil {
+				By("validating BFD parameters are propagated to the router")
+				for _, pod := range routerPods {
+					podExec := executor.ForPod(pod.Namespace, pod.Name, "frr")
```

Comment:
why does this check the config on podExec instead of checking it was accepted on the ToR (using the fields zapped from the struct)?